### ISSUE TITLE
Non-reproducibility of UMAP at larger data sizes

### ISSUE THREAD

**Issue by kevinthomasgo** [lmcinnes/umap#483]
Hello,

My question is: Does UMAP become unstable at larger data sizes for particular datasets?

I was running umap on [this dataset ](https://www.kaggle.com/roshansharma/online-shoppers-intention) on online shoppers.

I noticed that when running UMAP's `fit_transform()` on the **full dataset (~12k rows)**, I get different values of embeddings per run.

However, when I run `fit_transform()` on a **smaller sample of the dataset (~9k rows)**, UMAP produces stable embeddings.

Finally, when I do it on a **larger simulated dataset (20k rows)**, UMAP still produces stable embeddings.

Please see [this collab notebook](https://colab.research.google.com/drive/1kI1nmwBXLRJ3SfcCvRzKeloxR9oPaB7M?usp=sharing) for code and output. (If you want to run it for the online shoppers data, you'll need to upload the zip file of the dataset, and either run the notebook in Playground Mode or save a copy on your drive). 

Not sure why this happens for this particular dataset. Hope someone can help me.

Thank you!

**Comment by lmcinnes** [comment]
If you set a random seed it *should* be stable. If it isn't then there may be some quirkiness in that data that breaks things. One potential issue is duplicate points. If your data has a lot of duplicated points that can break some of the nearest neighbor estimation and resulting downstream things. A solution to that issue is to use the ``unique=True`` flag which will embed the unique points and then drop the duplicates down on top of their unique representative. That may help.

**Comment by kevinthomasgo** [comment]
Oh good catch, it was the duplicates causing it. After dropping them things were stable. Thanks!
